### PR TITLE
added the function and tests

### PR DESCRIPTION
**Pull request by mikecann**
As discussed in: https://github.com/kentcdodds/nps-utils/issues/22

For this PR, I support the happy-path if you are using the standard yarn workspaces or lerna structure:

```
module.exports = {
  scripts: {
    client: importPackage("client"),
    server: importPackage("server"),
    shared: importPackage("shared"),
  }
}
```

Also support you optionally providing the path:

```
module.exports = {
  scripts: {
    client: importPackage({path: "./mycustomfolder/my-nps-scripts.js" })
  }
}
```

It detects "descriptions" and handles them correctly.

Wasnt too sure how to structure the required test files (client, server). 

Fixed most of the es-lint issues but one, no arrow functions?

Plus i'm not sure how to actually functionally test this other than using console.log, perhaps you can advise?

### CODE REVIEW

**Comment by mikecann** [comment]
Probably this should respect the `options` object of the imported script too

File: src/__snapshots__/index.test.js.snap [hunk 1]
```diff
@@ -28,6 +28,14 @@ exports[`ifWindows as darwin 1`] = `"echo alternate"`;
 
 exports[`ifWindows as win32 1`] = `"echo main"`;
 
+exports[`includePackage as darwin 1`] = `"{\\"root\\":\\"cd packages/client && npm start root\\",\\"nested\\":{\\"child1\\":\\"cd packages/client && npm start nested.child1\\",\\"child2\\":{\\"script\\":\\"cd packages/client && npm start nested.child2 && cd /\\"C://dev//me//experiments//nps-utils/\\"\\",\\"description\\":\\"this is a description\\"},\\"child3\\":{\\"deep\\":{\\"script\\":\\"cd packages/client && npm start nested.child3.deep && cd /\\"C://dev//me//experiments//nps-utils/\\"\\",\\"description\\":\\"this is a description\\"}}}}"`;
```
Comment (kentcdodds) [inline src/__snapshots__/index.test.js.snap]: Looks like you're on a windows machine, so that's why the snapshots are failing. We need to serialize it differently to make sure things are working properly regardless of the platform. That's what `relativeizePath` is supposed to do. Could you figure out why it's not doing that?

Actually... Now that I think of it, I think this is a bit of a different kind of feature so I don't think I want to take a snapshot of it like we do with the others. Why don't you add a `test` block at the bottom of the file rather than include it in the `snapshotTests` object...

**Review by kentcdodds — COMMENTED** [review]
Looking good! Thanks for working on this! Let's try to make the tests work :+1:

File: src/__snapshots__/index.test.js.snap (hunk 1 above)
Comment (mikecann) [reply src/__snapshots__/index.test.js.snap]: Ye I thought so. It doesn't really seem to fit the snapshots quite so much. Ill write some tests for it instead.

Comment (mikecann) [reply src/__snapshots__/index.test.js.snap]: Im struggling with testing this as im not sure how to mock "require" correctly.

```import {includePackage} from './index'

test('it runs the default on a package', () => {
  const packageName = 'foo'
  jest.mock(`./packages/${packageName}/package-scripts.js`, () => ({
    scripts: {
      default: 'echo hello',
    },
  }))
  const outp = includePackage(packageName)
  expect(outp).toBe('{"default": "cd packages/foo && npm start default"')
})
```

Any tips?

Comment (kentcdodds) [reply src/__snapshots__/index.test.js.snap]: Hi @mikecann, don't put a `jest.mock` call inside a `test`. It doesn't work like that. Put it at the top of the file. Also, I'm not entirely sure how things work with mocking files that don't actually exist... You may have to make a `__fixtures__` directory and put real files in there.

I'm sorry. I have very limited time to give to maintaining this package :-/ Good luck.

File: src/index.js [hunk 2]
```diff
@@ -326,6 +327,67 @@ function crossEnv(args) {
...
+  const startingDir = process.cwd()
+  const relativeDir = path.relative(startingDir,
+    path.dirname(packageScriptsPath)).split('\\').join('/')
+  const relativeReturn = path.relative(relativeDir, startingDir)
+
+  let scripts = null
+  try {
+    scripts = require(packageScriptsPath)
+  } catch (e) {
+    throw Error(commonTags.oneLine`Couldnt include the package,
```
Comment (kentcdodds) [inline src/index.js]: This is helpful to have a good error message, but let's only throw it if we're sure the reason the original error was thrown is because the file does not exist (maybe do a require.resolve first or check the type of error). I've been bitten by this in the past where the error thrown was actually a syntax error in the file and I spent a lot of time trying to figure out why it couldn't find my file when it turned out that it could and the error was a red herring.

File: src/index.js [hunk 3]
```diff
@@ -326,6 +327,67 @@ function crossEnv(args) {
...
+ * @return {any} will return an object of scripts loaded from that package
+ */
+function includePackage(packageNameOrOptions) {
+  const packageScriptsPath = typeof packageNameOrOptions === 'string' ?
+    `./packages/${packageNameOrOptions}/package-scripts.js` :
+    packageNameOrOptions.path
+  
+  const startingDir = process.cwd()
+  const relativeDir = path.relative(startingDir,
+    path.dirname(packageScriptsPath)).split('\\').join('/')
```
Comment (kentcdodds) [inline src/index.js]: Will this work the same on Windows and Linux?

File: src/index.js [hunk 4]
```diff
@@ -326,6 +327,67 @@ function crossEnv(args) {
...
+      } else {
+        retObj[key] = Object.assign({}, replace(
+            obj[key], `${prefix}${dot}${key}`, `cd "${startingDir}"`))
+      }
+    }
+    return retObj
+  }
+ 
+  const outp = replace(scripts.scripts, '')
+  //console.log(JSON.stringify(outp,null,2));
```
Comment (kentcdodds) [inline src/index.js]: Let's remove this 👍

File: src/index.test.js [hunk 5]
```diff
@@ -55,6 +55,11 @@ const snapshotTests = {
     setColors(['white.bgBlue.bold', 'black.bgYellow.dim'])
     return concurrent.nps('lint', 'build')
   },
+  // includePackage: ({includePackage}) =>
+  //  JSON.stringify(includePackage('client')),
+  // includePackageWithPath: ({includePackage}) =>
+  //   JSON.stringify(includePackage({path:
+  //     './packages/someotherstructure/server/some-other-scripts-name.js'})),
```
Comment (kentcdodds) [inline src/index.test.js]: Let's remove these

File: src/index.test.js [hunk 6]
```diff
@@ -68,6 +73,8 @@ Object.keys(snapshotTests).forEach(testName => {
   })
 })
 
+
```
Comment (kentcdodds) [inline src/index.test.js:71]: Let's remove these extra lines

File: src/packages/client/package-scripts.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+module.exports = {
```
Comment (kentcdodds) [inline src/packages/client/package-scripts.js]: Is this file necessary?

File: src/packages/someotherstructure/server/some-other-scripts-name.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
```
Comment (kentcdodds) [inline src/packages/someotherstructure/server/some-other-scripts-name.js]: I'd this file necessary?

Comment (kentcdodds) [reply src/packages/someotherstructure/server/some-other-scripts-name.js]: If these files are necessary, let's add a comment to them indicating why they exist 👍

File: src/includePackage.test.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
+import {includePackage, commonTags} from './index'
+
+test('given just a string, it looks for scripts at a default location', () => {
+
+  jest.mock(`./packages/foo/package-scripts.js`, () => ({
+    scripts: {
+      default: 'echo foo',
+    },
+  }), {virtual: true})
```
Comment (kentcdodds) [inline src/includePackage.test.js:9]: Oooohhhh! `virtual: true` is something new to me! Cool! So does this mean you can mock files that don't exist? This is very useful. TIL!

**Review by kentcdodds — COMMENTED** [review]
This pull request keeps getting better! Great work! Just a few things.

**Comment by kentcdodds** [comment]
Would you like to [add yourself](../blob/master/CONTRIBUTING.md#add-yourself-as-a-contributor) to the [contributors table](../#contributors)?

File: src/includePackage.test.js (hunk 9 above)
Comment (mikecann) [reply src/includePackage.test.js:9]: Yep! However a gotcha is they seem to stay alive for the entire lifetime of all the tests not just the scope of that one test. Hence why I was careful to have a unique module path each time I used it.

File: src/index.js (hunk 3 above)
Comment (mikecann) [reply src/index.js]: Ah you know what, I think thats why I was having travis issues. Ill look at it now.

File: src/index.js (hunk 2 above)
Comment (mikecann) [reply src/index.js]: Oh thats a very good point. Ill write some tests for syntax errors too.

Comment (mikecann) [reply src/index.js]: Okay unfortunately require.resolve doesnt work with jest mocks, so decided that the extra complication probably isnt worth it and the native errors are good enough.

File: src/index.js (hunk 4 above)
Comment (mikecann) [reply src/index.js]: oops, yes, sorry

File: src/index.test.js (hunk 6 above)
Comment (mikecann) [reply src/index.test.js:71]: yep sorry, not a good citizen

File: src/packages/client/package-scripts.js (hunk 7 above)
Comment (mikecann) [reply src/packages/client/package-scripts.js]: nope, removed

File: src/packages/someotherstructure/server/some-other-scripts-name.js (hunk 8 above)
Comment (mikecann) [reply src/packages/someotherstructure/server/some-other-scripts-name.js]: nope, sorry they arent needed any more

**Comment by mikecann** [comment]
@kentcdodds thank you so much for the review, really helpful. I have fixed everything that I could. The only thing is those helpful error messages had to go due to limitations in jest mocking, the native errors are okay.

I have updated the readme to add the new method, I hope is is okay. Im not sure how the docs get generated so couldnt check that they look okay.

I ran into issues automatically testing this on multiple platforms, a simple solution was to replace '\\' with '/' which seems to work on windows just fine these days.

**Comment by kentcdodds** [comment]
This is great. Thank you so much for iterating and teaching me through this process!

**Comment by mikecann** [comment]
@kentcdodds the pleasure was all mine!

**Comment by kentcdodds** [comment]
Thanks so much for your help! I've added you as a collaborator on the project. Please make sure that you review the `other/MAINTAINING.md` and `CONTRIBUTING.md` files (specifically the bit about the commit messages and the git hooks) and familiarize yourself with the code of conduct (we're using [the contributor covenant](http://contributor-covenant.org/)). You might also want to [watch the repo](https://help.github.com/articles/watching-repositories/) to be notified when someone files an issue/PR. Please continue to make PRs as you feel the need (you can make your branches directly on the repo rather than your fork if you want). Thanks! And welcome to the team :)

**Comment by mikecann** [comment]
Thats awesome thanks Kent. Just one thing, do the docs auto-generate after some time?

**Comment by kentcdodds** [comment]
Well... It's supposed to, but it never really worked well. It looks like things have been updated on the esdoc site and it seems to be working mostly, but it doesn't have your new method. I'm not sure why not. Why don't we give it a day and see. If it doesn't update then maybe we should just put docs in the README...

**Comment by mikecann** [comment]
Sure, sounds like a plan, perhaps they have some cronjob to check for updates.
